sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py

```diff
@@ -160,37 +161,40 @@ class BlobClient(StorageAccountHostsMixin, StorageEncryptionMixin):  # pylint: d
             :caption: Creating the BlobClient from a SAS URL to a blob.
     """
     def __init__(
-            self, account_url: str,
-            container_name: str,
-            blob_name: str,
-            snapshot: Optional[Union[str, Dict[str, Any]]] = None,
-            credential: Optional[Union[str, Dict[str, str], "AzureNamedKeyCredential", "AzureSasCredential", "TokenCredential"]] = None,  # pylint: disable=line-too-long
-            **kwargs: Any
-        ) -> None:
+        self, account_url: str,
+        container_name: str,
+        blob_name: str,
+        snapshot: Optional[Union[str, Dict[str, Any]]] = None,
+        credential: Optional[Union[str, Dict[str, str], "AzureNamedKeyCredential", "AzureSasCredential", "TokenCredential"]] = None,  # pylint: disable=line-too-long
+        **kwargs: Any
+    ) -> None:
         parsed_url, sas_token, path_snapshot = _parse_url(
             account_url=account_url,
             container_name=container_name,
             blob_name=blob_name)
         self.container_name = container_name
         self.blob_name = blob_name
-        try:
-            self.snapshot = snapshot.snapshot # type: ignore
-        except AttributeError:
-            try:
-                self.snapshot = snapshot['snapshot'] # type: ignore
-            except TypeError:
-                self.snapshot = snapshot or path_snapshot
+
+        if snapshot is None:
+            self.snapshot = snapshot or path_snapshot
```

Comment:
nit: We know `snapshot` is `None` so the or-statement will always pick `path_snapshot` so probably can just go with that.

That said does MyPy require us to the do the None check first? I think it would look nicer to be

```py
if hasattr(snapshot, 'snapshot'):
    self.snapshot = snapshot.snapshot
elif isinstance(snapshot, dict):
    if 'snapshot' in snapshot:
        self.snapshot = snapshot['snapshot']
else:
    self.snapshot = snapshot or path_snapshot
```

Reply:
Additionally for the dictionary one can we change back to `self.snapshot = snapshot['snapshot']`? This will throw an error if the element is not in the dictionary which I think we want so customers don't think we accepted their snapshot when we didn't.